### PR TITLE
Cover parts of `go_modules` code with Sorbet

### PR DESCRIPTION
**Pull request by ByAgenT**
This PR adds `strict` Sorbet annotation in some files in `go_modules` and enables Sorbet Runtime.

### CODE REVIEW

File: go_modules/lib/dependabot/go_modules/file_fetcher.rb [hunk 1]
```diff
@@ -37,23 +40,21 @@ def fetch_files
...
       end
 
       private
 
+      sig { returns(Dependabot::DependencyFile) }
       def go_mod
-        return @go_mod if defined?(@go_mod)
-
-        @go_mod = fetch_file_if_present("go.mod")
+        @go_mod ||= T.let(T.must(fetch_file_if_present("go.mod")), T.nilable(Dependabot::DependencyFile))
```
Comment (ByAgenT) [inline go_modules/lib/dependabot/go_modules/file_fetcher.rb:52]: @JamieMagee can you please confidence check me if this code is equivalent to what was before?

Comment (JamieMagee) [reply go_modules/lib/dependabot/go_modules/file_fetcher.rb:52]: The `if_present` suffix on the end of this makes is sounds like `nil` might actually be a valid value for `@go_mod`. So the `T.must` may be unnecessary here.

Comment (JamieMagee) [reply go_modules/lib/dependabot/go_modules/file_fetcher.rb:52]: Just checked, and the test failure is due to the `T.must` here.

File: go_modules/lib/dependabot/go_modules/file_fetcher.rb [hunk 2]
```diff
@@ -35,25 +38,23 @@ def fetch_files
           directory,
           clone_repo_contents
         ) do
-          fetched_files = [go_mod]
+          fetched_files = go_mod ? [go_mod] : []
```
Comment (ByAgenT) [inline go_modules/lib/dependabot/go_modules/file_fetcher.rb:41]: Added conditional here to match the type

File: go_modules/lib/dependabot/go_modules/file_fetcher.rb (hunk 1 above)
Comment (ByAgenT) [reply go_modules/lib/dependabot/go_modules/file_fetcher.rb]: Thanks, I figured out this call tree by adding `nil` check to `fetch_files` method, tests are passing now

File: go_modules/lib/dependabot/go_modules/file_parser.rb [hunk 3]
```diff
@@ -85,56 +95,63 @@ def required_packages
 
             stdout, stderr, status = Open3.capture3(command)
             handle_parser_error(path, stderr) unless status.success?
-            JSON.parse(stdout)["Require"] || []
-          end
+            T.let(JSON.parse(stdout)["Require"] || [], T::Array[T::Hash[String, T.untyped]])
```
Comment (JamieMagee) [inline go_modules/lib/dependabot/go_modules/file_parser.rb:98]: Is this necessary? Or is it also caught by the outer `T.let`?

**Review by JamieMagee — APPROVED** [review]

File: go_modules/lib/dependabot/go_modules/file_parser.rb (hunk 3 above)
Comment (ByAgenT) [reply go_modules/lib/dependabot/go_modules/file_parser.rb]: Good catch, doesn't look like it's necessary!

**Review by abdulapopoola — APPROVED** [review]

File: go_modules/lib/dependabot/go_modules/file_fetcher.rb (hunk 2 above)
Comment (ryanbrandenburg) [inline go_modules/lib/dependabot/go_modules/file_fetcher.rb:41]: Oh ruby DOES have ternary expressions, dono why I couldn't find that.

**Review by ryanbrandenburg — APPROVED** [review]

**Pull request merged**